File: android2/app/src/main/java/com/xbot/xbot/voice/PophieClient.java
package com.xbot.xbot.voice;

import android.util.Base64;
import android.util.Log;

import androidx.annotation.Nullable;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import com.google.gson.reflect.TypeToken;

import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.reflect.Type;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;
import okio.ByteString;

/**
 * HTTP/WebSocket client for the Pophie backend.
 */
public class PophieClient {
    private static final String TAG = "PophieClient";
    private static final Gson GSON = new Gson();
    private static final Type MAP_TYPE = new TypeToken<Map<String, Object>>() {}.getType();
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    private final OkHttpClient http;
    private PophieConfig config;

    public PophieClient() {
        this(new PophieConfig());
    }

    public PophieClient(PophieConfig config) {
        this.config = config;
        this.http = new OkHttpClient.Builder()
                .connectTimeout(15, TimeUnit.SECONDS)
                .readTimeout(60, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .build();
    }

    public PophieConfig getConfig() {
        return config;
    }

    public void setConfig(PophieConfig config) {
        this.config = config;
    }

    @Nullable
    public String getSessionId() {
        return config.sessionId;
    }

    public void setSessionId(@Nullable String sessionId) {
        config.sessionId = sessionId;
    }

    /** Multimodal perception context (no object-detection fields). */
    public static class PophiePerception {
        @Nullable public String facialExpression;
        @Nullable public String identity;
        @Nullable public String gestureType;
        @Nullable public String touch;
        @Nullable public String scene;

        public boolean isEmpty() {
            return isBlank(facialExpression)
                    && isBlank(identity)
                    && isBlank(gestureType)
                    && isBlank(touch)
                    && isBlank(scene);
        }

        public Map<String, Object> toJson() {
            Map<String, Object> m = new HashMap<>();
            if (!isBlank(facialExpression)) {
                m.put("facial_expression", facialExpression);
            }
            if (!isBlank(identity)) {
                m.put("identity", identity);
            }
            if (!isBlank(gestureType)) {
                Map<String, Object> gesture = new HashMap<>();
                gesture.put("type", gestureType);
                m.put("gesture", gesture);
            }
            if (!isBlank(touch)) {
                m.put("touch", touch);
            }
            if (!isBlank(scene)) {
                m.put("scene", scene);
            }
            return m;
        }

        private static boolean isBlank(@Nullable String s) {
            return s == null || s.trim().isEmpty();
        }
    }

    public static class PophieChatResult {
        public final String text;
        public final String facialExpression;
        public final String robotState;
        public final String sttText;
        @Nullable public final byte[] audioBytes;
        @Nullable public final String audioFormat;
        @Nullable public final Map<String, Object> voice;
        @Nullable public final String sessionId;

        public PophieChatResult(
                String text,
                String facialExpression,
                String robotState,
                String sttText,
                @Nullable byte[] audioBytes,
                @Nullable String audioFormat,
                @Nullable Map<String, Object> voice,
                @Nullable String sessionId) {
            this.text = text != null ? text : "";
            this.facialExpression = facialExpression != null ? facialExpression : "neutral";
            this.robotState = robotState != null ? robotState : "idle";
            this.sttText = sttText != null ? sttText : "";
            this.audioBytes = audioBytes;
            this.audioFormat = audioFormat;
            this.voice = voice;
            this.sessionId = sessionId;
        }

        public boolean isSilent() {
            return text.trim().isEmpty();
        }
    }

    public enum SttEventType {
        META, READY, PARTIAL, FINAL_RESULT, SESSION_END, ERROR, CLOSED
    }

    public static class SttEvent {
        public final SttEventType type;
        @Nullable public final String text;
        @Nullable public final Map<String, Object> voice;
        @Nullable public final String message;
        @Nullable public final Integer silenceCommitMs;
        @Nullable public final Integer conversationIdleSec;

        public SttEvent(SttEventType type) {
            this(type, null, null, null, null, null);
        }

        public SttEvent(
                SttEventType type,
                @Nullable String text,
                @Nullable Map<String, Object> voice,
                @Nullable String message,
                @Nullable Integer silenceCommitMs,
                @Nullable Integer conversationIdleSec) {
            this.type = type;
            this.text = text;
            this.voice = voice;
            this.message = message;
            this.silenceCommitMs = silenceCommitMs;
            this.conversationIdleSec = conversationIdleSec;
        }
    }

    public interface SttEventListener {
        void onEvent(SttEvent event);
    }

    public interface TtsStreamCallbacks {
        void onMeta(String format, int sampleRate);

        void onChunk(byte[] pcm);

        void onDone(@Nullable Integer firstPacketMs);

        void onError(String message);
    }

    /** POST /api/chat */
    public PophieChatResult chat(
            @Nullable String text,
            @Nullable byte[] wavBytes,
            @Nullable PophiePerception perception,
            @Nullable String userId,
            @Nullable Boolean skipTts) throws IOException {
        Map<String, Object> input = buildInput(text, wavBytes, perception, skipTts);
        Map<String, Object> body = buildChatBody(input, userId);

        Request request = new Request.Builder()
                .url(trimBaseUrl(config.baseUrl) + "/api/chat")
                .post(RequestBody.create(GSON.toJson(body), JSON))
                .build();

        try (Response response = http.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("POST /api/chat failed: HTTP " + response.code());
            }
            ResponseBody responseBody = response.body();
            if (responseBody == null) {
                throw new IOException("POST /api/chat empty body");
            }
            Map<String, Object> data = parseMap(responseBody.string());
            if (data == null) {
                throw new IOException("POST /api/chat invalid JSON");
            }
            Object sid = data.get("session_id");
            if (sid instanceof String) {
                config.sessionId = (String) sid;
            }
            Map<String, Object> output = asMap(data.get("output"));
            Map<String, Object> stt = asMap(data.get("stt"));
            byte[] audioBytes = null;
            String audioFormat = null;
            Map<String, Object> audio = asMap(output.get("audio"));
            if (audio != null && audio.get("data") instanceof String) {
                audioBytes = Base64.decode((String) audio.get("data"), Base64.DEFAULT);
                Object fmt = audio.get("format");
                audioFormat = fmt instanceof String ? (String) fmt : "wav";
            }
            Map<String, Object> voice = asMap(output.get("voice"));
            return new PophieChatResult(
                    stringVal(output.get("text")),
                    stringVal(output.get("facial_expression")),
                    stringVal(output.get("robot_state")),
                    stt != null ? stringVal(stt.get("text")) : "",
                    audioBytes,
                    audioFormat,
                    voice,
                    config.sessionId);
        }
    }

    /** Streaming TTS: POST /api/tts/stream (NDJSON). */
    public void ttsStream(
            String text,
            @Nullable Map<String, Object> voice,
            @Nullable String voiceId,
            TtsStreamCallbacks callbacks) throws IOException {
        Map<String, Object> body = new HashMap<>();
        body.put("text", text);
        String vid = voiceId != null ? voiceId : config.voiceId;
        if (vid != null && !vid.isEmpty()) {
            body.put("voice_id", vid);
        }
        if (voice != null && !voice.isEmpty()) {
            body.put("voice", voice);
        }

        Request request = new Request.Builder()
                .url(trimBaseUrl(config.baseUrl) + "/api/tts/stream")
                .post(RequestBody.create(GSON.toJson(body), JSON))
                .header("Accept", "application/x-ndjson")
                .build();

        try (Response response = http.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("POST /api/tts/stream failed: HTTP " + response.code());
            }
            ResponseBody responseBody = response.body();
            if (responseBody == null) {
                throw new IOException("POST /api/tts/stream empty body");
            }
            BufferedReader reader = new BufferedReader(
                    new InputStreamReader(responseBody.byteStream(), StandardCharsets.UTF_8));
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty()) {
                    continue;
                }
                handleTtsNdjsonLine(line, callbacks);
            }
        }
    }

    private void handleTtsNdjsonLine(String line, TtsStreamCallbacks callbacks) throws IOException {
        Map<String, Object> obj = parseMap(line);
        if (obj == null) {
            return;
        }
        String type = stringVal(obj.get("type"));
        switch (type) {
            case "meta":
                callbacks.onMeta(
                        stringVal(obj.get("format"), "pcm"),
                        intVal(obj.get("sample_rate"), 22050));
                break;
            case "chunk":
                Object data = obj.get("data");
                if (data instanceof String && !((String) data).isEmpty()) {
                    callbacks.onChunk(Base64.decode((String) data, Base64.DEFAULT));
                }
                break;
            case "done":
                callbacks.onDone(intVal(obj.get("first_packet_ms"), null));
                break;
            case "error":
                String msg = stringVal(obj.get("message"), "未知错误");
                callbacks.onError(msg);
                throw new IOException("/api/tts/stream 合成失败: " + msg);
            default:
                Log.w(TAG, "ttsStream unknown type: " + type);
                break;
        }
    }

    /** Open WebSocket STT session at /api/stt/stream. */
    public SttStreamSession openSttStream(SttEventListener listener) {
        String url = toWsUrl(config.baseUrl);
        SttStreamSession session = new SttStreamSession(listener);
        Request request = new Request.Builder().url(url).build();
        WebSocket ws = http.newWebSocket(request, session);
        session.attach(ws);
        return session;
    }

    public static String toWsUrl(String baseUrl) {
        String b = trimBaseUrl(baseUrl);
        if (b.startsWith("https://")) {
            return "wss://" + b.substring("https://".length()) + "/api/stt/stream";
        }
        if (b.startsWith("http://")) {
            return "ws://" + b.substring("http://".length()) + "/api/stt/stream";
        }
        return "ws://" + b + "/api/stt/stream";
    }

    public static byte[] pcm16ToWav(byte[] pcm) {
        return pcm16ToWav(pcm, 16000, 1, 16);
    }

    public static byte[] pcm16ToWav(byte[] pcm, int sampleRate, int channels, int bitsPerSample) {
        int byteRate = sampleRate * channels * bitsPerSample / 8;
        int blockAlign = channels * bitsPerSample / 8;
        int dataLen = pcm.length;
        ByteBuffer header = ByteBuffer.allocate(44).order(ByteOrder.LITTLE_ENDIAN);
        header.put("RIFF".getBytes(StandardCharsets.US_ASCII));
        header.putInt(36 + dataLen);
        header.put("WAVE".getBytes(StandardCharsets.US_ASCII));
        header.put("fmt ".getBytes(StandardCharsets.US_ASCII));
        header.putInt(16);
        header.putShort((short) 1);
        header.putShort((short) channels);
        header.putInt(sampleRate);
        header.putInt(byteRate);
        header.putShort((short) blockAlign);
        header.putShort((short) bitsPerSample);
        header.put("data".getBytes(StandardCharsets.US_ASCII));
        header.putInt(dataLen);
        ByteArrayOutputStream out = new ByteArrayOutputStream(44 + dataLen);
        out.write(header.array(), 0, 44);
        out.write(pcm, 0, dataLen);
        return out.toByteArray();
    }

    public void shutdown() {
        http.dispatcher().executorService().shutdown();
        http.connectionPool().evictAll();
    }

    private Map<String, Object> buildInput(
            @Nullable String text,
            @Nullable byte[] wavBytes,
            @Nullable PophiePerception perception,
            @Nullable Boolean skipTts) {
        Map<String, Object> input = new HashMap<>();
        input.put("text", text != null ? text : "");
        if (wavBytes != null && wavBytes.length > 0) {
            Map<String, Object> audio = new HashMap<>();
            audio.put("format", "wav");
            audio.put("encoding", "base64");
            audio.put("sample_rate", 16000);
            audio.put("data", Base64.encodeToString(wavBytes, Base64.NO_WRAP));
            input.put("audio", audio);
        }
        if (perception != null && !perception.isEmpty()) {
            input.put("perception", perception.toJson());
        }
        if (config.voiceId != null && !config.voiceId.isEmpty()) {
            input.put("voice_id", config.voiceId);
        }
        if (skipTts != null) {
            input.put("skip_tts", skipTts);
        }
        return input;
    }

    private Map<String, Object> buildChatBody(Map<String, Object> input, @Nullable String userId) {
        Map<String, Object> body = new HashMap<>();
        body.put("robot_id", config.robotId);
        if (config.sessionId != null) {
            body.put("session_id", config.sessionId);
        }
        if (userId != null && !userId.isEmpty()) {
            body.put("user_id", userId);
        }
        body.put("input", input);
        return body;
    }

    private static String trimBaseUrl(String baseUrl) {
        String b = baseUrl != null ? baseUrl.trim() : "";
        while (b.endsWith("/")) {
            b = b.substring(0, b.length() - 1);
        }
        return b;
    }

    @Nullable
    private static Map<String, Object> parseMap(String json) {
        try {
            return GSON.fromJson(json, MAP_TYPE);
        } catch (JsonSyntaxException e) {
            return null;
        }
    }

    @SuppressWarnings("unchecked")
    @Nullable
    private static Map<String, Object> asMap(@Nullable Object obj) {
        if (obj instanceof Map) {
            return (Map<String, Object>) obj;
        }
        return null;
    }

    private static String stringVal(@Nullable Object v) {
        return stringVal(v, "");
    }

    private static String stringVal(@Nullable Object v, String fallback) {
        return v instanceof String ? (String) v : fallback;
    }

    @Nullable
    private static Integer intVal(@Nullable Object v, @Nullable Integer fallback) {
        if (v instanceof Number) {
            return ((Number) v).intValue();
        }
        return fallback;
    }

    /** WebSocket client for streaming STT. */
    public class SttStreamSession extends WebSocketListener {
        private final SttEventListener listener;
        private final AtomicBoolean closed = new AtomicBoolean(false);
        @Nullable private WebSocket webSocket;

        SttStreamSession(SttEventListener listener) {
            this.listener = listener;
        }

        void attach(WebSocket webSocket) {
            this.webSocket = webSocket;
        }

        public void start(int sampleRate, String language, @Nullable Boolean turnDetection) {
            Map<String, Object> msg = new HashMap<>();
            msg.put("type", "start");
            msg.put("sample_rate", sampleRate);
            msg.put("language", language);
            if (turnDetection != null) {
                msg.put("turn_detection", turnDetection);
            }
            send(msg);
        }

        public void sendChunk(byte[] pcm16) {
            if (closed.get() || pcm16.length == 0) {
                return;
            }
            Map<String, Object> msg = new HashMap<>();
            msg.put("type", "chunk");
            msg.put("data", Base64.encodeToString(pcm16, Base64.NO_WRAP));
            send(msg);
        }

        public void commit() {
            send(Map.of("type", "commit"));
        }

        public void sendEnd() {
            send(Map.of("type", "end"));
        }

        public void close(boolean sendEndFrame) {
            if (!closed.compareAndSet(false, true)) {
                return;
            }
            if (sendEndFrame) {
                sendEnd();
            }
            if (webSocket != null) {
                webSocket.close(1000, "client close");
            }
            emit(new SttEvent(SttEventType.CLOSED));
        }

        private void send(Map<String, Object> msg) {
            if (closed.get() || webSocket == null) {
                return;
            }
            webSocket.send(GSON.toJson(msg));
        }

        private void emit(SttEvent event) {
            if (!closed.get() || event.type == SttEventType.CLOSED) {
                listener.onEvent(event);
            }
        }

        @Override
        public void onOpen(WebSocket webSocket, Response response) {
            Log.d(TAG, "STT WebSocket connected");
        }

        @Override
        public void onMessage(WebSocket webSocket, String text) {
            Map<String, Object> obj = parseMap(text);
            if (obj == null) {
                return;
            }
            String type = stringVal(obj.get("type"));
            switch (type) {
                case "meta":
                    emit(new SttEvent(
                            SttEventType.META,
                            null,
                            null,
                            null,
                            intVal(obj.get("silence_commit_ms"), null),
                            intVal(obj.get("conversation_idle_sec"), null)));
                    break;
                case "ready":
                    emit(new SttEvent(SttEventType.READY));
                    break;
                case "partial":
                    emit(new SttEvent(
                            SttEventType.PARTIAL,
                            stringVal(obj.get("text")),
                            null,
                            null,
                            null,
                            null));
                    break;
                case "final":
                    emit(new SttEvent(
                            SttEventType.FINAL_RESULT,
                            stringVal(obj.get("text")),
                            asMap(obj.get("voice")),
                            null,
                            null,
                            null));
                    break;
                case "session_end":
                    emit(new SttEvent(
                            SttEventType.SESSION_END,
                            null,
                            null,
                            stringVal(obj.get("message")),
                            null,
                            intVal(obj.get("conversation_idle_sec"), null)));
                    break;
                case "error":
                    emit(new SttEvent(
                            SttEventType.ERROR,
                            null,
                            null,
                            stringVal(obj.get("message")),
                            null,
                            null));
                    break;
                default:
                    Log.w(TAG, "STT unknown frame: " + type);
                    break;
            }
        }

        @Override
        public void onMessage(WebSocket webSocket, ByteString bytes) {
            // protocol uses text frames only
        }

        @Override
        public void onFailure(WebSocket webSocket, Throwable t, @Nullable Response response) {
            Log.e(TAG, "STT WebSocket failure", t);
            emit(new SttEvent(
                    SttEventType.ERROR,
                    null,
                    null,
                    t.getMessage(),
                    null,
                    null));
            closed.set(true);
            emit(new SttEvent(SttEventType.CLOSED));
        }

        @Override
        public void onClosed(WebSocket webSocket, int code, String reason) {
            closed.set(true);
            emit(new SttEvent(SttEventType.CLOSED));
        }
    }
}

File: xbot-native/app/src/main/java/com/xbot/android/voice/PophieClient.kt
package com.xbot.android.voice

import android.util.Base64
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Pophie 后端客户端（STT+LLM+TTS）。对应 Flutter pophie_client.dart。
 *
 * 端点：
 * - POST /api/chat          批量 STT+LLM+TTS（skip_tts=true 时只回文本）
 * - POST /api/chat/stream   流式对话 NDJSON：{type:speak,text} → {type:done,response}
 * - POST /api/tts/stream    流式 TTS NDJSON：meta → chunk(base64 pcm) → done
 * - GET  /api/proactive_messages   主动消息（每 5s 轮询）
 * - PUT/GET/DELETE /api/robots/{robot_id}/owner   主人档案
 *
 * 音频：16kHz mono PCM16 WAV（44 字节头）base64。TTS 返回 PCM16 mono @22050。
 */
class PophieClient(
    var config: PophieConfig,
) {
    companion object {
        private const val TAG = "PophieClient"
        private val JSON = "application/json; charset=utf-8".toMediaType()
        private val NDJSON = "application/x-ndjson".toMediaType()
    }

    val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(8, TimeUnit.SECONDS)
        .readTimeout(40, TimeUnit.SECONDS)
        .build()

    // ============ /api/chat（批量）============

    data class ChatResult(
        val text: String,
        val facialExpression: String?,
        val robotState: String?,
        val sttText: String?,
        val sessionId: String?,
        val isSilent: Boolean = text.isBlank(),
    )

    /**
     * 批量对话：发送 WAV(base64) + 感知上下文，skip_tts=true（流式 TTS 单独走）。
     * @param wavBytes 16kHz mono WAV；为 null 则纯文本输入
     * @param text 纯文本输入（voice-only 时为 ""）
     */
    suspend fun chat(
        wavBytes: ByteArray? = null,
        text: String = "",
        perception: JSONObject? = null,
        userId: String? = null,
    ): ChatResult = withContext(Dispatchers.IO) {
        val input = JSONObject()
        if (text.isNotEmpty()) input.put("text", text)
        if (wavBytes != null) {
            val audio = JSONObject()
            audio.put("format", "wav")
            audio.put("encoding", "base64")
            audio.put("sample_rate", 16000)
            audio.put("data", Base64.encodeToString(wavBytes, Base64.NO_WRAP))
            input.put("audio", audio)
        }
        if (perception != null) input.put("perception", perception)
        if (config.voiceId.isNotEmpty()) input.put("voice_id", config.voiceId)
        input.put("skip_tts", true)

        val body = JSONObject()
        body.put("robot_id", config.robotId)
        config.sessionId?.let { body.put("session_id", it) }
        userId?.let { body.put("user_id", it) }
        body.put("input", input)

        val req = Request.Builder()
            .url("${config.baseUrl.trimEnd('/')}/api/chat")
            .post(body.toString().toRequestBody(JSON))
            .build()
        val resp = await(client.newCall(req))
        val root = JSONObject(resp.body!!.string())
        val output = root.optJSONObject("output")
        val sid = root.optString("session_id")
        if (sid.isNotEmpty()) config.sessionId = sid
        ChatResult(
            text = output?.optString("text") ?: "",
            facialExpression = output?.optString("facial_expression"),
            robotState = output?.optString("robot_state"),
            sttText = root.optJSONObject("stt")?.optString("text"),
            sessionId = sid.takeIf { it.isNotEmpty() },
        )
    }

    // ============ /api/chat/stream（NDJSON）============

    /**
     * 流式对话：逐句 onSpeak（TTS 每段），最终 onDone 返回完整结果。
     */
    suspend fun chatStream(
        text: String,
        perception: JSONObject? = null,
        userId: String? = null,
        onSpeak: (String) -> Unit,
    ) = withContext(Dispatchers.IO) {
        val input = JSONObject()
        input.put("text", text)
        if (perception != null) input.put("perception", perception)
        if (config.voiceId.isNotEmpty()) input.put("voice_id", config.voiceId)
        input.put("skip_tts", true)
        val body = JSONObject()
        body.put("robot_id", config.robotId)
        config.sessionId?.let { body.put("session_id", it) }
        userId?.let { body.put("user_id", it) }
        body.put("input", input)

        val req = Request.Builder()
            .url("${config.baseUrl.trimEnd('/')}/api/chat/stream")
            .post(body.toString().toRequestBody(JSON))
            .header("Accept", "application/x-ndjson")
            .build()
        val resp = await(client.newCall(req))
        val reader = BufferedReader(InputStreamReader(resp.body!!.byteStream()))
        var result: ChatResult = ChatResult(text = "", null, null, null, null)
        var line = reader.readLine()
        while (line != null) {
            if (line.isNotBlank()) {
                try {
                    val obj = JSONObject(line)
                    when (obj.optString("type")) {
                        "speak" -> onSpeak(obj.optString("text"))
                        "done" -> {
                            val output = obj.optJSONObject("response")?.optJSONObject("output")
                            val sid = obj.optJSONObject("response")?.optString("session_id")
                            if (!sid.isNullOrEmpty()) config.sessionId = sid
                            result = ChatResult(
                                text = output?.optString("text") ?: "",
                                facialExpression = output?.optString("facial_expression"),
                                robotState = output?.optString("robot_state"),
                                sttText = obj.optJSONObject("response")?.optJSONObject("stt")?.optString("text"),
                                sessionId = sid,
                            )
                        }
                        "error" -> throw IOException(obj.optString("message"))
                    }
                } catch (_: Exception) {
                    // 跳过坏行。
                }
            }
            line = reader.readLine()
        }
        result
    }

    // ============ /api/tts/stream（NDJSON）============

    /**
     * 流式 TTS：meta(sampleRate) → chunk(base64 pcm16) → done。
     * @param onMeta 返回采样率（默认 22050）
     * @param onChunk 返回解码后的 PCM16 字节
     * @param onDone 返回首包延迟 ms
     */
    suspend fun ttsStream(
        text: String,
        voice: JSONObject? = null,
        onMeta: (Int) -> Unit,
        onChunk: (ByteArray) -> Unit,
        onDone: (Int) -> Unit,
    ) = withContext(Dispatchers.IO) {
        val body = JSONObject()
        body.put("text", text)
        if (voice != null) body.put("voice", voice)
        if (config.voiceId.isNotEmpty()) body.put("voice_id", config.voiceId)
        val req = Request.Builder()
            .url("${config.baseUrl.trimEnd('/')}/api/tts/stream")
            .post(body.toString().toRequestBody(JSON))
            .header("Accept", "application/x-ndjson")
            .build()
        val call = client.newCall(req)
        val resp = await(call)
        val reader = BufferedReader(InputStreamReader(resp.body!!.byteStream()))
        var line = reader.readLine()
        while (line != null) {
            if (line.isNotBlank()) {
                try {
                    val obj = JSONObject(line)
                    when (obj.optString("type")) {
                        "meta" -> onMeta(obj.optInt("sample_rate", 22050))
                        "chunk" -> {
                            val data = obj.optString("data")
                            if (data.isNotEmpty()) onChunk(Base64.decode(data, Base64.NO_WRAP))
                        }
                        "done" -> onDone(obj.optInt("first_packet_ms"))
                        "error" -> throw IOException(obj.optString("message"))
                    }
                } catch (e: IOException) {
                    throw e
                } catch (_: Exception) {
                    // 跳过坏行。
                }
            }
            line = reader.readLine()
        }
    }

    /** 取消所有进行中的请求（barge-in 用）。 */
    fun cancelAll() {
        client.dispatcher.cancelAll()
    }

    /** 健康检查（设置页「测试连接」）：能拿到任意 HTTP 响应即视为后端可达。 */
    suspend fun health(): Boolean = withContext(Dispatchers.IO) {
        try {
            val req = Request.Builder()
                .url("${config.baseUrl.trimEnd('/')}/api/schema")
                .get()
                .build()
            await(client.newCall(req)).use { it.code in 200..499 }
        } catch (_: Exception) {
            false
        }
    }

    // ============ /api/proactive_messages ============

    data class ProactiveMessage(val id: Long, val content: String, val trigger: String)

    suspend fun fetchProactiveMessages(sinceId: Long): List<ProactiveMessage> = withContext(Dispatchers.IO) {
        val url = "${config.baseUrl.trimEnd('/')}/api/proactive_messages" +
            "?robot_id=${config.robotId}&since_id=$sinceId&limit=10" +
            (config.sessionId?.let { "&session_id=$it" } ?: "")
        val req = Request.Builder().url(url).get().build()
        val resp = await(client.newCall(req))
        val root = JSONObject(resp.body!!.string())
        val items = root.optJSONArray("items") ?: JSONArray()
        val out = ArrayList<ProactiveMessage>(items.length())
        for (i in 0 until items.length()) {
            val it = items.getJSONObject(i)
            val meta = it.optJSONObject("metadata")
            val trigger = meta?.optString("trigger") ?: "reminder"
            out.add(ProactiveMessage(it.optLong("id"), it.optString("content"), trigger))
        }
        out
    }

    // ============ /api/robots/{id}/owner ============

    suspend fun registerOwner(ownerJson: String): Boolean = withContext(Dispatchers.IO) {
        val req = Request.Builder()
            .url("${config.baseUrl.trimEnd('/')}/api/robots/${config.robotId}/owner")
            .put(ownerJson.toRequestBody(JSON))
            .build()
        try { await(client.newCall(req)).code in 200..299 } catch (_: Exception) { false }
    }

    // ============ helpers ============

    /** OkHttp Call → Response（挂起，自动 close body 由调用方在读完流后处理）。 */
    private suspend fun await(call: Call): Response = suspendCancellableCoroutine { cont ->
        cont.invokeOnCancellation { runCatching { call.cancel() } }
        call.enqueue(object : Callback {
            override fun onFailure(c: Call, e: IOException) {
                Log.e(TAG, "请求失败: ${e.message}")
                if (cont.isActive) cont.resumeWithException(e)
            }
            override fun onResponse(c: Call, r: Response) {
                if (cont.isActive) cont.resume(r)
            }
        })
    }

    /** PCM16 → WAV（44 字节头 + PCM）。 */
    fun pcm16ToWav(pcm16: ShortArray): ByteArray {
        val byteCount = pcm16.size * 2
        val out = java.io.ByteArrayOutputStream(44 + byteCount)
        val little = java.nio.ByteOrder.LITTLE_ENDIAN
        val bb = java.nio.ByteBuffer.allocate(44).order(little)
        bb.put("RIFF".toByteArray()); bb.putInt(36 + byteCount); bb.put("WAVE".toByteArray())
        bb.put("fmt ".toByteArray()); bb.putInt(16); bb.putShort(1) // PCM
        bb.putShort(1); bb.putInt(16000); bb.putInt(16000 * 2) // byteRate
        bb.putShort(2); bb.putShort(16) // bits
        bb.put("data".toByteArray()); bb.putInt(byteCount)
        out.write(bb.array())
        val data = java.nio.ByteBuffer.allocate(byteCount).order(little)
        for (s in pcm16) data.putShort(s)
        out.write(data.array())
        return out.toByteArray()
    }

    /** PCM16 short[] → byte[]（小端）。 */
    fun pcm16ToBytes(samples: ShortArray): ByteArray {
        val bb = java.nio.ByteBuffer.allocate(samples.size * 2).order(java.nio.ByteOrder.LITTLE_ENDIAN)
        for (s in samples) bb.putShort(s)
        return bb.array()
    }
}
